Support read rate-limiting in SequentialFileReader

cbi42 opened the pull request:
Summary: Added rate limiter and read rate-limiting support to SequentialFileReader. I've updated call sites to SequentialFileReader::Read with appropriate IO priority (or left a TODO and specified IO_TOTAL for now). 

The PR is separated into four commits: the first one added the rate-limiting support, but with some fixes in the unit test since the number of request bytes from rate limiter in SequentialFileReader are not accurate (there is overcharge at EOF). The second commit fixed this by allowing SequentialFileReader to check file size and determine how many bytes are left in the file to read. The third commit added benchmark related code. The fourth commit moved the logic of using file size to avoid overcharging the rate limiter into backup engine (the main user of SequentialFileReader).

Test Plan: 
- `make check`, backup_engine_test covers usage of SequentialFileReader with rate limiter.
- Run db_bench to check if rate limiting is throttling as expected: Verified that reads and writes are together throttled at 2MB/s, and at 0.2MB chunks that are 100ms apart.
  - Set up: `./db_bench --benchmarks=fillrandom -db=/dev/shm/test_rocksdb`
  - Benchmark: 
```
strace -ttfe read,write ./db_bench --benchmarks=backup -db=/dev/shm/test_rocksdb --backup_rate_limit=2097152 --use_existing_db
strace -ttfe read,write ./db_bench --benchmarks=restore -db=/dev/shm/test_rocksdb --restore_rate_limit=2097152 --use_existing_db
```
- db bench on backup and restore to ensure no performance regression.
  - backup (avg over 50 runs): pre-change: 1.90443e+06 micros/op; post-change: 1.8993e+06 micros/op (improve by 0.2%)
  - restore (avg over 50 runs): pre-change: 1.79105e+06 micros/op; post-change: 1.78192e+06 micros/op (improve by 0.5%)


```
# Set up
./db_bench --benchmarks=fillrandom -db=/tmp/test_rocksdb -num=10000000

# benchmark
TEST_TMPDIR=/tmp/test_rocksdb
NUM_RUN=50
for ((j=0;j<$NUM_RUN;j++))
do
   ./db_bench -db=$TEST_TMPDIR -num=10000000 -benchmarks=backup -use_existing_db | egrep 'backup'
  # Restore
  #./db_bench -db=$TEST_TMPDIR -num=10000000 -benchmarks=restore -use_existing_db
done > rate_limit.txt && awk -v NUM_RUN=$NUM_RUN '{sum+=$3;sum_sqrt+=$3^2}END{print sum/NUM_RUN, sqrt(sum_sqrt/NUM_RUN-(sum/NUM_RUN)^2)}' rate_limit.txt >> rate_limit_2.txt
```

hx235 commented on utilities/backup/backup_engine_test.cc:
```diff
@@ -2696,7 +2703,6 @@ TEST_P(BackupEngineRateLimitingTestWithParam, RateLimitingVerifyBackup) {
   auto rate_limited_verify_backup_time =
       (bytes_read_during_verify_backup * kMicrosPerSec) /
       backup_rate_limiter_limit;
-
```
Potential merge conflict with https://github.com/facebook/rocksdb/pull/9974, depending on which one of us merges first - shouldn't be a tough one though

cbi42 replied:
Thanks for the heads up!

hx235 commented on file/sequence_file_reader.cc:
```diff
@@ -25,19 +26,32 @@ namespace ROCKSDB_NAMESPACE {
 }
 
-IOStatus SequentialFileReader::Read(size_t n, Slice* result, char* scratch) {
+IOStatus SequentialFileReader::Read(size_t n, Slice* result, char* scratch,
```
cc @anand1976  our file read wrapper expert - would you mind also taking a quick look into this PR's change on `SequentialFileReader::Read`.  Although I checked against its original impl as well as an example `RandomAccessFileReader::Read` PLUS [ReadaheadSequentialFileTest](https://github.com/facebook/rocksdb/blob/dd203ed6041000ba728a7ad92b3d568ccd4d4e21/util/file_reader_writer_test.cc#L586) seems to be a good testing coverage for (SequentialFileReader + no rate limiter) case, I'd still appreciate a quick look from domain expert. Thanks!

hx235 replied:
Reached out offline.

hx235 commented on file/sequence_file_reader.cc:
```diff
@@ -25,19 +26,32 @@ namespace ROCKSDB_NAMESPACE {
+                                    Env::IOPriority rate_limiter_priority) {
   IOStatus io_s;
+  uint64_t filesize;
+  IOOptions io_opts;
```
`IOOptions io_opts;` does not seem to be used by your current impl

cbi42 replied:
Good catch..not sure why. This may be the reason? https://stackoverflow.com/questions/22112784/why-is-unused-variable-warning-not-reported-for-all-variables

hx235 commented on file/sequence_file_reader.cc:
```diff
@@ -25,19 +26,32 @@ namespace ROCKSDB_NAMESPACE {
+    return io_s;
+  }
+  // Cap the bytes limit to number of remaining bytes in the file
+  n = std::min(n, static_cast<size_t>(filesize) - offset_);
```
Should it be ...offset_.load() since is a atomic var? I am surprised that type check does not catch anything for your current usage (i.e, "-" operation between a size_t and a atomic<size_t> 🤔

Another thing I'd do is to `assert(static_cast<size_t>(filesize) >= offset_.load())` (which only takes effects in non-prod env) as comparing a (theoretically possible but not practically possible as `offset_` should not be greater than the `filesize`, assuming both variables stay correct in impl) negative number with `size_t n` (non-negative) can give surprising casting behavior.

cbi42 replied:
Good catch! Is SequentialFileReader expected to be thread-safe btw?

ajkr replied:
> Should it be ...offset_.load() since it's an atomic var? I am surprised that type check does not catch anything for your current usage (i.e, "-" operation between a size_t and a atomic<size_t> 🤔

This usage does the same thing.

hx235 replied:
I was under impression that atomic_var does not have a `-` operation defined looking at https://en.cppreference.com/w/cpp/atomic/atomic so I assume this will need to go through some casting that isn't standard (atomic<size_t> -> size_t). 

I'd love to know the reference to "This usage does the same thing." as it's not obvious to me :p

cbi42 replied:
Maybe these two? https://en.cppreference.com/w/cpp/atomic/atomic/operator%3D
https://en.cppreference.com/w/cpp/atomic/atomic/operator_T

ajkr replied:
yeah probably implicit conversion (https://en.cppreference.com/w/cpp/atomic/atomic/operator_T) to pass to operator+(size_t, size_t). I don't plan to confirm it.

hx235 replied:
@cbi42 @ajkr  thanks - then both ways are good with me!

hx235 commented on file/sequence_file_reader.cc:
```diff
@@ -25,19 +26,32 @@ namespace ROCKSDB_NAMESPACE {
+  n = std::min(n, static_cast<size_t>(filesize) - offset_);
   if (use_direct_io()) {
 #ifndef ROCKSDB_LITE
+
```
Empty line not needed

hx235 commented on file/sequence_file_reader.cc:
```diff
@@ -48,53 +62,87 @@ IOStatus SequentialFileReader::Read(size_t n, Slice* result, char* scratch) {
+      if (rate_limiter_priority != Env::IO_TOTAL && rate_limiter_ != nullptr) {
+        allowed = rate_limiter_->RequestToken(
+            buf.Capacity() - buf.CurrentSize(), buf.Alignment(),
+            rate_limiter_priority, nullptr, RateLimiter::OpType::kRead);
```
nit: add comment nullptr like /* stats */ - ditto for other RequestToken() below in this PR.

cbi42 replied:
sg, added.

hx235 commented on file/sequence_file_reader.cc:
```diff
@@ -48,53 +62,87 @@ IOStatus SequentialFileReader::Read(size_t n, Slice* result, char* scratch) {
+    if (io_s.ok() && offset_advance < buf.CurrentSize()) {
       r = buf.Read(scratch, offset_advance,
-                   std::min(tmp.size() - offset_advance, n));
+                   std::min(buf.CurrentSize() - offset_advance, n));
```
Not relevant to @cbi42 's impl but just for my learning:

cc @anand1976 our file read wrapper expert - not relevant to this PR's impl but it seems to me that 
(1) We are not returning the first `offset_advance` bytes in the `AlignedBuf buf` by `r = buf.Read(scratch, offset_advance,...)` to user's result. Are those bytes garbage? I understand that we need to read aligned but don't understand why we abandon those bytes.

(2) It also seems to me that `offset_advance < buf.CurrentSize()` is trying to guard from the **only** edge case where `offset_advance == buf.CurrentSize()`, which is the case **only where** `n==0` and offset is right at some alignment (i.e, `offset % alignment = 0`). This further implies `offset_advance == 0` and `buf.CurrentSize() == 0`.

Am I understanding this correctly?

cbi42 replied:
From my understanding:
(1) The first `offset_advance` bytes are the bytes from `aligned_offset` to `offset`, which are probably already read by the caller in previous calls, and the caller is expecting to read sequentially for bytes after `offset`, so these `offset_advance` bytes before `offset` are discarded.
(2) I think it is possible to have `offset_advance > buf.CurrentSize()`, e.g., when the call to `file_->PositionedRead` returns ok status but `tmp.size() < offset_advance`. Not sure if that's likely tho.
If we consider pre-PR when file size is not considered, then if `Read` is called when `offset` and `aligned_offset` are already at end of file, then I think we could have offset_advance > buf.CurrentSize() too.

hx235 replied:
@cbi42 aha - thanks! Seems like we could use some clarification like this https://github.com/facebook/rocksdb/pull/10002 :)

hx235 commented on file/sequence_file_reader.cc:
```diff
@@ -48,53 +62,87 @@ IOStatus SequentialFileReader::Read(size_t n, Slice* result, char* scratch) {
-      size_t offset = offset_.fetch_add(result->size());
-      NotifyOnFileReadFinish(offset, result->size(), start_ts, finish_ts, io_s);
-    }
+      size_t offset = offset_.fetch_add(tmp.size());
```
This line should go inside of `if (ShouldNotifyListeners()){}` since no reference to it outside of this if block

cbi42 replied:
Updated.

hx235 commented:
Not finished reviewing yet

hx235 commented on file/line_file_reader.h:
```diff
@@ -41,7 +43,8 @@ class LineFileReader {
   // end-of-file (OK status) or an I/O error (another status).
-  bool ReadLine(std::string* out);
+  bool ReadLine(std::string* out,
+                Env::IOPriority rate_limiter_priority = Env::IO_TOTAL);
```
Would you mind adding a comment about `rate_limiter_priority` since there exists function comment for `out`

hx235 commented on utilities/backup/backup_engine.cc:
```diff
@@ -2039,7 +2040,8 @@ IOStatus BackupEngineImpl::CopyOrCreateFile(
     *src_temperature = src_file->GetTemperature();
 
-    src_reader.reset(new SequentialFileReader(std::move(src_file), src));
+    src_reader.reset(new SequentialFileReader(std::move(src_file), src, nullptr,
```
Could you add comment on this nullptr like`\* io_tracer *\` - ditto for LineFileReader in https://github.com/facebook/rocksdb/pull/9973/files?show-viewed-files=true&file-filters%5B%5D=#diff-e9c5eba5c364f5a2a4164c928c29121c8bcdaf4dbea6541517f625554f196076R25

hx235 commented on utilities/backup/backup_engine.cc:
```diff
@@ -2079,14 +2077,9 @@ IOStatus BackupEngineImpl::CopyOrCreateFile(
     io_s = dest_writer->Append(data);
 
     if (rate_limiter != nullptr) {
-      if (!src.empty()) {
-        rate_limiter->Request(data.size(), Env::IO_LOW, nullptr /* stats */,
-                              RateLimiter::OpType::kWrite);
-      } else {
```
Good catch 👍- not migrating OpType::kWrite in backup engine makes it easy to overlook this needed change.

hx235 commented on utilities/backup/backup_engine.cc:
```diff
@@ -2356,16 +2349,14 @@ IOStatus BackupEngineImpl::ReadFileAndComputeChecksum(
-  IOStatus io_s = SequentialFileReader::Create(src_fs, src, file_options,
-                                               &src_reader, nullptr);
+  RateLimiter* rate_limiter = options_.backup_rate_limiter.get();
+  IOStatus io_s = SequentialFileReader::Create(
```
Not relevant to @cbi42 's impl: for our conversation whether RandomAccessFileReader is used in BackupEngine,  [sst_reader](https://github.com/facebook/rocksdb/blob/0d8433c5853e468975382b05b92cdd72379c64c8/utilities/backup/backup_engine.cc#L2391) [does wrap it and use it .](https://github.com/facebook/rocksdb/blob/main/table/sst_file_dumper.h#L90). 

Though this is not relevant to this PR as it only migrates SequentialFileReader part

cbi42 replied:
Should we have a task to track improving `RandomAccessFileReader` with file size?

hx235 replied:
See the follow-up task :p

hx235 commented on utilities/backup/backup_engine.cc:
```diff
@@ -2079,14 +2077,9 @@ IOStatus BackupEngineImpl::CopyOrCreateFile(
 
     if (rate_limiter != nullptr) {
-      if (!src.empty()) {
-        rate_limiter->Request(data.size(), Env::IO_LOW, nullptr /* stats */,
```
I believe in the case of `if (!src.empty())` we don't need to loop as we still have `size_t buf_size = rate_limiter ? static_cast<size_t>(rate_limiter->GetSingleBurstBytes()): kDefaultCopyFileBufferSize;` [above](https://github.com/facebook/rocksdb/pull/9973/files?show-viewed-files=true&file-filters%5B%5D=#diff-4d85a375a1fe9d14d0e5cfa18bf19729e614a4f3edf73ca59b0e7bde0b1e79c6L2030-L2032)  AND in the case of `!src.empty()`, we are writing what we read and the constrained `buf_size` still takes effect as it is passed down in `io_s = src_reader->Read(buffer_to_read, &data, buf.get(), Env::IO_LOW)` [here](https://github.com/facebook/rocksdb/pull/9973/files?show-viewed-files=true&file-filters%5B%5D=#diff-4d85a375a1fe9d14d0e5cfa18bf19729e614a4f3edf73ca59b0e7bde0b1e79c6R2056). So the `buffer_to_read` >= data.size() that we are going to write.


Unlike the else case  (i.e, `src.empty() ==  true`) where we are writing whatever is passed into this function (i.e, `data = contents;` in https://github.com/facebook/rocksdb/pull/9973/files?show-viewed-files=true&file-filters%5B%5D=#diff-4d85a375a1fe9d14d0e5cfa18bf19729e614a4f3edf73ca59b0e7bde0b1e79c6L2061) so the constrained buffer_to_read does not take effect.


However, using `LoopRateLimitRequestHelper` for both case like the change here should be fine as the `LoopRateLimitRequestHelper` just iterate once for the former case.

hx235 replied:
[No action needed from author and just a note]

I think having `size_t buf_size = rate_limiter ? static_cast<size_t>(rate_limiter->GetSingleBurstBytes()): kDefaultCopyFileBufferSize;` actually limits the exercise of the newly added lower-level rate-limiting in SequentialFileReader as the single read `io_s = src_reader->Read(buffer_to_read...` is already within burst.  

But `size_t buf_size = rate_limiter ? static_cast<size_t>(rate_limiter->GetSingleBurstBytes()): kDefaultCopyFileBufferSize;` is still needed as we haven't migrated the write rate-limiting. Remove it while not migrating write will leads to `io_s = dest_writer->Append(data);` [here](https://github.com/facebook/rocksdb/blob/0d8433c5853e468975382b05b92cdd72379c64c8/utilities/backup/backup_engine.cc#L2077) actually write-more-than-burst before rate-limit.

hx235 commented on include/rocksdb/env.h:
```diff
@@ -738,6 +739,9 @@ class SequentialFile {
   }
 
+  virtual Status GetFileSize(uint64_t& /* size */) const {
+    return Status::NotSupported("GetFileSize");
```
nit: I'd stay aligned with how existing `Status::NotSupported` msgs above are - `return Status::NotSupported("SequentialFile::GetFileSize not supported.");`

cbi42 replied:
Ah okay, copy paste mistake :)

hx235 commented on file/sequence_file_reader.cc:
```diff
@@ -25,19 +26,32 @@ namespace ROCKSDB_NAMESPACE {
+  // file after last read
+  io_s = file_->GetFileSize(filesize);
+  if (!io_s.ok()) {
+    return io_s;
```
cc @anand1976, @cbi42 - I wonder if we should fail the read just because an optional API `GetFileSize` is not implemented. Plus it is only needed when there's a need for rate-limiting. Even when we actually need rate-limiting, is this a strong enough reason to fail read?

An alternative is to fall back to rate-limiting not knowing the file size (so no `n = std::min(n, static_cast<size_t>(filesize) - offset_);` below, which leads to some extra unnecessary rate-limit request) but does not fail read.

hx235 replied:
I don't feel "io_s = file_->GetFileSize(filesize);" will slow down `SequentialFileReader::Read` much but I'd still suggest:
(1) Move it into if-branch when rate-limiting is needed as it's only used for rate-limiting purpose so far
(2) db bench or microbench on backup/restore for pre and (post PR + instrumentally mocking the actual wait in rate-limiter (e.g, temporarily remove https://github.com/facebook/rocksdb/blob/7.2.fb/util/rate_limiter.cc#L38 )). The reason to mock out the actual wait in rate-limiter is that we only interested in knowing how much `io_s = file_->GetFileSize(filesize);` will slow down read. 

Although I don't anticipate regression from (2) especially with the optimization you are exploring right now, I'd still stay cautious and run some benchmark.

For more stable db_bench running, you can leverage a recent feaeture like https://github.com/facebook/rocksdb/pull/9882

cbi42 replied:
@anand1976 
> No one should be simultaneously reading and writing the file.

I found [here](https://github.com/facebook/rocksdb/blob/c4cd8e1accfd010bba9a919e234ba04505071cc1/db/log_reader.h#L94 
) that log_reader has this `UnmarkEof` function, and it will read from a `SequentialFileReader` if it knows that more data is appended to the file. Do you think that's a legit use case?

Regarding `GetFileSize() need not be called for every read. I'm working to improve this with @hx235's suggestion above and only refresh file size when reading beyond end of file.

> we should call fs->GetFileSize() to get the size rather than add a new API in FSSequentialFile

Good point, thanks, I'll try this change.

anand1976 replied:
> I found here that log_reader has this UnmarkEof function, and it will read from a SequentialFileReader if it knows that more data is appended to the file. Do you think that's a legit use case?

That's a good point. Maybe the secondary instance functionality uses it to tail the MANIFEST, among other things. @riversand963 is the expert in that.

ajkr replied:
I suspect we currently do not require `GetFileSize()` even once per file. It might be doing that on local filesystem if you strace a backup, but when run on an Env/FileSystem that implements GetChildrenFileAttributes() it should not. The motivation is, on distributed file systems like HDFS, querying file size goes through the namenode, which can be (and has been at times) the scalability bottleneck.

So I guess the question is, is there a way to provide file size that backup/checkpoint already got from `GetChildrenFileAttributes()` to as the initial value for `SequentialFileReader`'s size?

cbi42 replied:
If we can get file size from `GetChildrenFileAttributes()`, this would save the initial call. However, we still need another call when reading beyond the last byte to check if new content was added to the file, so it will still be once per file?

ajkr replied:
Oh ok. Maybe just try to prevent regression in a common existing use case (backup local flash <-> remote FS). In that case the restore operation is the relevant one because in that case we are reading files from remote FS. I think we can prevent reading past EOF by passing `file_info->size` as `size_limit` here https://github.com/facebook/rocksdb/blob/f6339de0d2e0383faa90fdf011a665c94be8f692/utilities/backup/backup_engine.cc#L1925

ajkr replied:
I wasn't following what test failures this GetFileSize() intends to address. Have you considered the alternative of restricting how much bytes clients of SequentialFileReader call for Read()? Looks like CopyFile() already does this (based on its own GetFileSize()), and BackupEngine could too as it also knows file size. Are there more places similar logic would be needed to fix the test failures? Or, are there other places we are worried about over-charging?

cbi42 replied:
@ajkr I have not tried the alternative, and I think limiting the size should work for CopyFile() and BackupEngine. The worry is for over-charging in log reader, as it reads a log file that may be written to at the same time. Maybe that's fine (currently I used IO_TOTAL for it), and I can add a TODO for fixing that overcharge later.

cbi42 replied:
To summarize, I have updated the PR to restrict BackupEngine (the main user of SequentialFileReader) on how many bytes it calls Read() for (so no additional system call is introduced). Specifically, I updated the `size_limit` parameter [here](https://github.com/facebook/rocksdb/blob/987ca52c9985afb3c085ed5cd272921dfc1165aa/utilities/backup/backup_engine.cc#L1436) and [here](https://github.com/facebook/rocksdb/blob/987ca52c9985afb3c085ed5cd272921dfc1165aa/utilities/backup/backup_engine.cc#L1861). 
I have added TODOs in `log_reader` about that it might overcharge `SequentialFileReader`'s rate limiter and will leave it for future work (it currently uses `IO_TOTAL` priority so this PR should not affect it). 
Thanks for the review and suggestions!

hx235 commented on tools/db_bench_tool.cc:
```diff
@@ -8185,6 +8207,50 @@ class Benchmark {
+  void Backup(ThreadState* thread) {
+    DB* db = SelectDB(thread);
+    std::string backup_dir = FLAGS_db + "/backups";
+    std::unique_ptr<BackupEngineOptions> engine_options_(
```
Underscore suffix like in `engine_options_` is usually for private data member (ref: "Data members of classes (but not structs) additionally have trailing underscores." in https://google.github.io/styleguide/cppguide.html#Variable_Names )- we can use `engine_options` here.

cbi42 replied:
Updated.

hx235 commented on tools/db_bench_tool.cc:
```diff
@@ -8185,6 +8207,50 @@ class Benchmark {
+    Status s;
+    BackupEngine* backup_engine;
+    if (FLAGS_backup_rate_limit > 0) {
+      engine_options_.get()->backup_rate_limiter.reset(NewGenericRateLimiter(
```
Is there a particular reason why we need the raw pointer of engine_options_ `engine_options_.get()` but not directly with unique_ptr?

cbi42 replied:
Good catch, no particular reason: IDE being weird and me being rusty in C++ ..

cbi42 replied:
Updated to use unique_ptr directly.

hx235 commented on tools/db_bench_tool.cc:
```diff
@@ -8185,6 +8207,50 @@ class Benchmark {
+          10 /* fairness */, RateLimiter::Mode::kAllIo));
+    }
+    BackupEngineReadOnly* backup_engine;
+    Status s = BackupEngineReadOnly::Open(Env::Default(), *engine_options_,
```
Should we use FLAGS_env instead of Env::Default() like backup above?

hx235 commented on tools/db_bench_tool.cc:
```diff
@@ -8185,6 +8207,50 @@ class Benchmark {
+                                          &backup_engine);
+    assert(s.ok());
+    std::string restore_dir =
+        FLAGS_db + "/restore_" + std::to_string(thread->rand.Next());
```
Any particular reason to add some randomness (i.e, thread->rand.Next())  to the restore folder name?

hx235 commented on tools/db_bench_tool.cc:
```diff
@@ -8185,6 +8207,50 @@ class Benchmark {
+                                          &backup_engine);
+    assert(s.ok());
+    std::string restore_dir =
+        FLAGS_db + "/restore_" + std::to_string(thread->rand.Next());
```
Is there any particular reason why we add randomness to the folder naming other than avoid collision across runs of `Benchmark::Restore`? 

I am under the impression that even when we pass the same restore folder name every time we call this `Benchmark::Restore` function, the same folder's content will be deleted first due to 
https://github.com/facebook/rocksdb/blob/7.2.fb/utilities/backup/backup_engine.cc#L1798-L1800.

cbi42 replied:
I had some issues that seem to be resolved now. Updated the code with randomness removed.

hx235 commented on tools/db_bench_tool.cc:
```diff
@@ -8185,6 +8207,50 @@ class Benchmark {
+  }
+
+  void Restore(ThreadState* thread) {
+    std::string backup_dir = FLAGS_db + "/backups";
```
Since we hard-coded the backup path, might worth printing out the restore/backup path like `DB path` so that people could know the folder name without looking into the code. Ditto for backup benchmark above.

cbi42 replied:
Added FLAGS_backup_dir and FLAGS_restore_dir, and the print statements.

hx235 commented on env/io_posix.cc:
```diff
@@ -320,6 +322,13 @@ IOStatus PosixSequentialFile::InvalidateCache(size_t offset, size_t length) {
+IOStatus PosixSequentialFile::GetFileSize(uint64_t& size) const {
+  IOOptions io_opts;
+  IOStatus io_s =
+      FileSystem::Default()->GetFileSize(filename_, io_opts, &size, nullptr);
```
cc @anand1976  - do you think it's the best way to GetFileSize (same for win case  [port/win/io_win.cc](https://github.com/facebook/rocksdb/pull/9973/files?show-viewed-files=true&file-filters%5B%5D=#diff-d461daf22c23b03cc251e07a31e89974fae6d1b7424056eeffa21de1e833dd55)). 

It looks right to me but I'm not sure.

cbi42 replied:
Resolve for now as this part of code is removed.

hx235 commented:
@cbi42 Great progress so far! I finished my round of review and left some comments - most are minor coding details excepts for https://github.com/facebook/rocksdb/pull/9973#discussion_r872868708, https://github.com/facebook/rocksdb/pull/9973#discussion_r872893885,https://github.com/facebook/rocksdb/pull/9973#discussion_r872900151,https://github.com/facebook/rocksdb/pull/9973#discussion_r872900921.


There are two places (https://github.com/facebook/rocksdb/pull/9973#discussion_r872867230, https://github.com/facebook/rocksdb/pull/9973#discussion_r872904884) where I'd like extra pair of eyes from domain expert @anand1976, where I left my comments. @cbi42  - would you mind following up with him on these particular two places? Thanks!

hx235 reviewed (commented):
@cbi42 Great progress so far!!! I finished current round of review and left some comments - most are minor coding details excepts for https://github.com/facebook/rocksdb/pull/9973#discussion_r872868708, https://github.com/facebook/rocksdb/pull/9973#discussion_r872893885, https://github.com/facebook/rocksdb/pull/9973#discussion_r872900151, https://github.com/facebook/rocksdb/pull/9973#discussion_r872900921.

There are two places (https://github.com/facebook/rocksdb/pull/9973#discussion_r872867230, https://github.com/facebook/rocksdb/pull/9973#discussion_r872904884) where I'd like an extra pair of eyes from domain expert @anand1976 and left my comments. @cbi42 - would you mind following up with him on these particular two places?

Further exploration on "refreshing file size in SequentialFileReader when requesting to read beyond the last byte" sounds great with me!

Thanks!

cbi42 commented on file/line_file_reader.h:
```diff
@@ -32,7 +33,8 @@ class LineFileReader {
                          std::unique_ptr<LineFileReader>* reader,
-                         IODebugContext* dbg);
+                         IODebugContext* dbg,
+                         RateLimiter* rate_limiter = nullptr);
```
Removed the default value and all call sites are updated.

hx235 commented on db/log_reader.cc:
```diff
@@ -303,8 +303,11 @@ void Reader::UnmarkEOFInternal() {
   Slice read_buffer;
+  // TODO: rate limit log reader
+  // Note that this might overcharge SequentialFileReader's internal rate
+  // limiter if priority is not IO_TOTAL (when we reach EOF).
```
nit: "... is not IO_TOTAL (when we reach EOF) due to implementation limitations" so that at lest someone (you or me) remembers there's left to do for this.

cbi42 replied:
Updated with a more clear TODO.

hx235 commented on file/line_file_reader.h:
```diff
@@ -41,7 +41,11 @@ class LineFileReader {
   // end-of-file (OK status) or an I/O error (another status).
-  bool ReadLine(std::string* out);
+  // Caller can set rate limiting priority other than the default value
+  // Env::IO_TOTAL and the internal rate will be charged at the specified
```
".. and internal rate limiter will .."? instead of "...rate..."

cbi42 replied:
Thanks for catching this.

hx235 commented on tools/db_bench_tool.cc:
```diff
@@ -8185,6 +8213,49 @@ class Benchmark {
   }
 
+  void Backup(ThreadState* thread) {
+    std::cout << "Backup path: [" << FLAGS_backup_dir << "]" << std::endl;
```
Is there a way to move the printing outside of the benchmarked function - detto for Restore() below? 

I am thinking of maybe something before https://github.com/facebook/rocksdb/blob/7.2.fb/tools/db_bench_tool.cc#L3516-L3518

cbi42 replied:
Do we want to print them even when backup/restore benchmarks are not ran? I currently put them inside Backup()/Restore() so that they don't get print out in other benchmarks. 

I do realize that I should at least print Backup Path **and** Restore Path for the restore benchmark tho.

hx235 replied:
Can we conditionally print the path based on the `name` of the benchmark being run? I see an example like this https://github.com/facebook/rocksdb/blob/7.2.fb/tools/db_bench_tool.cc#L3451

cbi42 replied:
Sg, updated.

hx235 commented on util/file_reader_writer_test.cc:
```diff
@@ -597,7 +597,8 @@ class ReadaheadSequentialFileTest : public testing::Test,
     Slice result;
-    Status s = test_read_holder_->Read(n, &result, scratch_.get());
+    Status s =
+        test_read_holder_->Read(n, &result, scratch_.get(), Env::IO_TOTAL);
```
nit: test_read_holder_->Read(n, &result, scratch_.get(), Env::IO_TOTAL /* rate_limiter_priority*/);

hx235 commented on file/sequence_file_reader.h:
```diff
@@ -91,12 +96,19 @@ class SequentialFileReader {
+  // `rate_limiter_priority` is used to charge the internal rate limiter when
+  // enabled. The special value `Env::IO_TOTAL` makes this operation bypass the
+  // rate limiter. The amount charged to the internal rate limiter is n, even
+  // when less than n bytes are read (e.g. at end of file). To avoid
```
".. even when less than n bytes are *actually* read (e.g, at end of file) .." Adding *actually* to differentiate the "read" in your comment than the `IOStatus Read(size_t n, Slice* result, char* scratch, Env::IOPriority rate_limiter_priority);` here a bit.

hx235 commented on file/sequence_file_reader.cc:
```diff
@@ -25,16 +25,18 @@ namespace ROCKSDB_NAMESPACE {
+                                    Env::IOPriority rate_limiter_priority) {
   IOStatus io_s;
   if (use_direct_io()) {
 #ifndef ROCKSDB_LITE
```
potential merge conflict with https://github.com/facebook/rocksdb/pull/10002

cbi42 replied:
This seems to not cause conflict so far.

hx235 replied:
It's a bit weird that this PR does not pick up https://github.com/facebook/rocksdb/pull/10002 (merged in 
upstream/main) yet not showing in Github UI any conflict. Anyway, it should automatically show up somewhere before you land plus that PR is not a code change.

hx235 commented on file/sequence_file_reader.cc:
```diff
@@ -48,53 +50,89 @@ IOStatus SequentialFileReader::Read(size_t n, Slice* result, char* scratch) {
-    }
+      if (ShouldNotifyListeners()) {
+        size_t offset = offset_.fetch_add(tmp.size());
+        auto finish_ts = FileOperationInfo::FinishNow();
```
I forgot to point out in my first-round review: is there any reason why we swap the order of `FileOperationInfo::FinishNow();` and `size_t offset = offset_.fetch_add(tmp.size());`?

cbi42 replied:
I edited that part due to come compilation error before, but forgot that I re-ordered the lines. It's swapped back now.

hx235 commented on file/sequence_file_reader.cc:
```diff
@@ -48,53 +50,89 @@ IOStatus SequentialFileReader::Read(size_t n, Slice* result, char* scratch) {
+        NotifyOnFileReadFinish(offset, tmp.size(), start_ts, finish_ts, io_s);
+      }
+#else
+      offset_.fetch_add(tmp.size());
```
Also forgot to point out in my first-round review: offset_ seems to only matter in #ifndef ROCKSDB_LITE so I don't think adding #else branch is needed here.

hx235 commented on file/sequence_file_reader.cc:
```diff
@@ -48,53 +50,89 @@ IOStatus SequentialFileReader::Read(size_t n, Slice* result, char* scratch) {
+        NotifyOnFileReadFinish(offset, tmp.size(), start_ts, finish_ts, io_s);
+      }
+#else
+      offset_.fetch_add(tmp.size());
```
@anand1976  - based on the existing code (i.e, before this PR), `offset_` is only used under  (a) "use_direct_io() == true" && `ifndef ROCKSDB_LITE` and (b) "use_direct_io() == false" && `ifndef ROCKSDB_LITE` && `if (ShouldNotifyListeners())`

We should probably: update its comment `std::atomic<size_t> offset_{0};  // read offset` to be more specific `// read offset for if direct_io is used` AND comment on the case (b) why we need`if (ShouldNotifyListeners())`

hx235 replied:
@cbi42  I don't think `offset_` matter in not ROCKSDB_LITE mode - so this macro #else branch is not needed?

cbi42 replied:
Yeah..actually, I'll put the definition of `offset_` in only non ROCKSDB_LITE code too.

hx235 replied:
For this PR, it's enough to remove the added macro else branch. You can move the definition of offset_ (along with updating the comment for it a bit - see my comment above) in another PR (i.e, a follow-up, not urgent PR). Thanks!

cbi42 replied:
Sg, removed the #else branch.

hx235 reviewed (commented):
~~I will resume my review focusing on the new things in backup engine by EOD.~~

ajkr commented on utilities/backup/backup_engine.cc:
```diff
@@ -2877,11 +2850,6 @@ IOStatus BackupEngineImpl::BackupMeta::LoadFromFile(
   std::vector<std::shared_ptr<FileInfo>> files;
   bool footer_present = false;
   while (backup_meta_reader->ReadLine(&line)) {
```
Should `Env::IO_LOW` be passed to all the `ReadLine()`s, especially this one?

Also if we're going to have `ReadLine()`'s priority argument have default value `Env::IO_TOTAL`, should the `ReadLine()` code have a TODO to finish migrating clients?

hx235 replied:
(I had no idea how I overlooked this as a reviewer, probably my eyes keep copying "https://github.com/facebook/rocksdb/pull/9973/files#diff-4d85a375a1fe9d14d0e5cfa18bf19729e614a4f3edf73ca59b0e7bde0b1e79c6R2790")

hx235 commented on utilities/backup/backup_engine.cc:
```diff
@@ -1432,7 +1433,7 @@ IOStatus BackupEngineImpl::CreateNewBackupWithMetadata(
               false /* shared */, "" /* src_dir */, fname,
               EnvOptions() /* src_env_options */, rate_limiter, type,
-              contents.size(), db_options.statistics.get(), 0 /* size_limit */,
+              contents.size(), db_options.statistics.get(), contents.size(),
```
I doubt whether this change is needed for this PR for the following reasons:
(1) backup_engine_test passes with reverting this change 
(2) Based on the semantics, `ReadFileAndComputeChecksum` seems to fall under `AddBackupFileWorkItem` of `copy_file_cb` path (https://github.com/facebook/rocksdb/pull/9973/files#diff-4d85a375a1fe9d14d0e5cfa18bf19729e614a4f3edf73ca59b0e7bde0b1e79c6R1428) instead of `create_file_cb` path (this).  So the change here does not seem to affect the `ReadFileAndComputeChecksum` 
(3) Instead, the [one](https://github.com/facebook/rocksdb/blob/d110e17ffb2c1f25990108de814717f99a329a8b/utilities/backup/backup_engine.cc#L1978) under `VerifyBackup` might be more concerning in terms of over-charging, although the current test does not reveal that. It's probably due to currently the test that accidentally revealed overcharging was written (by me) for cases involving both read and write rate-limiting while `VerifyBackup` is a pure read read-limiting.

But tracing through the test `TEST_P(BackupEngineRateLimitingTestWithParam, RateLimitingVerifyBackup)` + sum up the charged bytes `buffer_to_read` (https://github.com/facebook/rocksdb/blob/d110e17ffb2c1f25990108de814717f99a329a8b/utilities/backup/backup_engine.cc#L2374) should reveal that. 

If we really want to make sure your improvement to prevent overcharging is really helping, then maybe we should add some unit tests to compare "the exact charged request bytes (i.e, RateLimiter::GetTotalBytesThrough())" VS "the estimated actual read bytes" (how to get this info in test is a bit tricky - maybe Env::GetFileSize()?). I expect to see a relatively big gap between this two before your improvement and a smaller gap (if not exactly equal) after your improvement.

For (3), let me double check @cbi42  - is `TEST_P(BackupEngineRateLimitingTestWithParam, RateLimitingChargeReadInRestore)` the only test that failed without your improvement, which I believe can be solved by changing the test to compare `GetTotalRequests` instead of `GetTotalBytesThrough`? If that's the case, I wouldn't mind saving your over-charging improvement into a new PR with unit-test and gets the tests passing first for this PR (considering it's already big). @ajkr  - do you have an opinion on my suggestion?

Btw, @pdillinger can confirm on my understanding about BackupEngine in (2)/(3) - some quick context for you Peter: @cbi42 is working on making `BackupEngine` cap `SequentialFileReader::Read()` by file size so that the rate-limiting made inside SequentialFileReader by this PR does not charge more than file size (i.e, the over-charging problem). And `SequentialFileReader::Read()`  is called in `BackupEngineImpl::ReadFileAndComputeChecksum` and `BackupEngineImpl::CopyOrCreateFile`

hx235 replied:
Btw, @pdillinger can confirm on my understanding about BackupEngine in (2)/(3) - some quick context for you Peter: @cbi42 is working on making BackupEngine cap SequentialFileReader::Read() by file size so that the rate-limiting made inside SequentialFileReader by this PR does not charge more than file size (i.e, the over-charging problem). And SequentialFileReader::Read() is called in BackupEngineImpl::ReadFileAndComputeChecksum and BackupEngineImpl::CopyOrCreateFile

hx235 commented on utilities/backup/backup_engine.cc:
```diff
@@ -1383,6 +1383,7 @@ IOStatus BackupEngineImpl::CreateNewBackupWithMetadata(
           uint64_t size_bytes = 0;
           IOStatus io_st;
           if (type == kTableFile || type == kBlobFile) {
+            // TODO: can we just use size_limit_bytes and remove this call?
```
My guess is that due to [this feature](https://github.com/facebook/rocksdb/blob/7.2.fb/include/rocksdb/metadata.h#L66-L69), `size_limit` that passed in (e.g, [here](https://github.com/facebook/rocksdb/blob/main/utilities/checkpoint/checkpoint_impl.cc#L268)) can be smaller than the actual file size, which might not be what the proceeding steps want - @pdillinger might know more.

(By the way, I usually put confirmed TODO in code and put my "in-question" as a PR comment to reviewers, although I do see a couple of "in-question" TODO in the code base and I don't know why ....)

cbi42 replied:
Ah right, I remember asking @ajkr about this too, he mentioned that `size_limit` limits how many bytes to copy from WAL/MANIFEST. However, I thought SST files and Blob files are immutable so their size is fixed and not trimmed (as they are not "live" files)?

ajkr replied:
Yes, `size_limit` may be smaller for mutable files (WAL and MANIFEST) because those files can continue to be appended after we have taken a snapshot of live files, but we do not want those newer appends in the checkpoint/backup. Otherwise the MANIFEST might reference SST files not included in the backup/checkpoint, for example.

For SST files and blob files we should not ever limit to a portion of the full file size. I did not know we even populate `size_limit` for these files until you told me, but since we do it needs to be >= the file size.

cbi42 replied:
I've removed the TODO comment for now and we can continue the discussion here or offline.

hx235 commented on utilities/backup/backup_engine.cc:
```diff
@@ -1432,7 +1433,7 @@ IOStatus BackupEngineImpl::CreateNewBackupWithMetadata(
               false /* shared */, "" /* src_dir */, fname,
               EnvOptions() /* src_env_options */, rate_limiter, type,
-              contents.size(), db_options.statistics.get(), 0 /* size_limit */,
+              contents.size(), db_options.statistics.get(), contents.size(),
```
I doubt whether this change is needed for this PR for the following reasons:
(1) `backup_engine_test.cc` passes with reverting this change 
(2) Based on code search, I believe `create_file_cb` is only used for [creating the CURRENT file ](https://github.com/facebook/rocksdb/blob/7.2.fb/utilities/checkpoint/checkpoint_impl.cc#L237-L247) by writing the `info.replacement_contents` with no `SequentialFileReader:Read()` involved. So it goes down [the "Create" path ](https://github.com/facebook/rocksdb/blob/7.2.fb/utilities/backup/backup_engine.cc#L2059)in "CopyAndCreateFile".
(3) Based on the semantics, `create_file_cb` does not seem to call `ReadFileAndComputeChecksum` as it does not involving reading a file hence no checksum verification is needed.  So the change here does not seem to affect the `ReadFileAndComputeChecksum`
(4) Instead, the [one](https://github.com/facebook/rocksdb/blob/d110e17ffb2c1f25990108de814717f99a329a8b/utilities/backup/backup_engine.cc#L1978) under VerifyBackup might be more concerning in terms of over-charging, although the current test does not reveal that. It's probably due to the current test that accidentally revealed overcharging problem was written (by me) for cases involving both read and write rate-limiting while `VerifyBackup` is pure read read-limiting.

But tracing through the test `TEST_P(BackupEngineRateLimitingTestWithParam, RateLimitingVerifyBackup)` + sum up the charged [bytes buffer_to_read](https://github.com/facebook/rocksdb/blob/d110e17ffb2c1f25990108de814717f99a329a8b/utilities/backup/backup_engine.cc#L2374) should reveal that. 

If we really want to make sure your improvement to prevent overcharging is helping, then maybe we should add some unit tests to compare "the requested bytes (i.e, RateLimiter::GetTotalBytesThrough())" VS "the estimated actual read bytes" (how to get this info in test is a bit tricky - maybe Env::GetFileSize()?). I expect to see a relatively big gap between this two before your improvement and a smaller gap (if not exactly equal) after your improvement.

For (4), let me double check with @cbi42 : is `TEST_P(BackupEngineRateLimitingTestWithParam, RateLimitingChargeReadInRestore)` the only test failed without your improvement, which I believe can be solved by changing the test to compare `GetTotalRequests` instead of `GetTotalBytesThrough`? If that's the case, I wouldn't mind saving your over-charging improvement for a separate PR with unit-test and getting the tests passed for this PR (considering it's already big). 

@ajkr  - do you have an opinion on my suggestion above?

hx235 replied:
Btw, @pdillinger can confirm on my understanding about BackupEngine in (2)/(3).
 
Some quick context for you Peter: @cbi42 is working on making BackupEngine cap SequentialFileReader::Read() by file size so that the rate-limiting made inside SequentialFileReader by this PR does not charge more than file size (i.e, the over-charging problem). 

And SequentialFileReader::Read() is called in BackupEngineImpl::ReadFileAndComputeChecksum and BackupEngineImpl::CopyOrCreateFile

cbi42 replied:
@hx235 Yes BackupEngineRateLimitingTestWithParam.RateLimitingChargeReadInRestore is the only test that's failing before [this line](https://github.com/facebook/rocksdb/blob/d110e17ffb2c1f25990108de814717f99a329a8b/utilities/backup/backup_engine.cc#L1861) was updated with `file_info->size`.

You are right that `contents.size()` here is not needed, at least for passing all the current test cases. I've reverted the change here and will follow up with a separate PR with some new unit tests.

ajkr replied:
> @ajkr - do you have any opinion on my suggestions above?

 sounds good to me. I am glad if we move the rate limiting to the lower layer.

The code looks good to me. Will defer to Hui to accept since she knows the most detail.

hx235 commented on utilities/backup/backup_engine.cc:
```diff
@@ -1857,7 +1858,7 @@ IOStatus BackupEngineImpl::RestoreDBFromBackup(
         file_info->temp, "" /* contents */, backup_env_, db_env_,
         EnvOptions() /* src_env_options */, options_.sync,
-        options_.restore_rate_limiter.get(), 0 /* size_limit */,
+        options_.restore_rate_limiter.get(), file_info->size,
```
This seems fine with me.

cbi42 replied:
I've kept this change so that no change is needed for the current unit test (`BackupEngineRateLimitingTestWithParam.RateLimitingChargeReadInRestore`) and at least there's no rate limiter overcharge on the restore side.

hx235 replied:
I reached out to @pdillinger on his opinion on change like this. The change is fine here but he pointed out "there's a pre-existing missed opportunity for sanity checking. If I'm reading the code correctly, it will only read up to size_limit but does not check (in applicable cases - almost all) that the file size is as expected at read time. I will create a techdebt task for this.".


Feel free to reach out to him on the techdebt task if you are interested and available (you don't have to if you have other stuff lined up)

cbi42 replied:
That make sense, thanks for checking!

hx235 reviewed (commented):
Just finished reviewing the improvement to over-charging problem in BackupEngine and left some comments. I personally think it worths a separate PR with its own unit test and probably  involving Peter to review that as it's deeper into BackupEngine stuff that I anticipate. 

See my suggestion here https://github.com/facebook/rocksdb/pull/9973#discussion_r877804676.

Thanks for the hard work so far!

hx235 commented:
Note: you might want to rebase early cuz the recently merged changes https://github.com/facebook/rocksdb/blob/main/file/sequence_file_reader.cc#L41-L47, https://github.com/facebook/rocksdb/blob/main/utilities/backup/backup_engine_test.cc#L2599-L2748 don't seemed to be included in your PR nor does CI here showing any conflict. That feels weird to me - is your upstream set correctly or fetched recently?

```
/data/users/huixiao/rocksdb$ git rebase --onto upstream/main fddc62a9b459cd8d1eb2dfad8f12217e70ab109a^
Auto-merging utilities/backup/backup_engine_test.cc
CONFLICT (content): Merge conflict in utilities/backup/backup_engine_test.cc
Auto-merging utilities/backup/backup_engine.cc
Auto-merging util/file_reader_writer_test.cc
Auto-merging file/sequence_file_reader.cc
Auto-merging db/repair.cc
error: could not apply fddc62a9b... Added RateLimiter support in SequentialFileReader
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply fddc62a9b... Added RateLimiter support in SequentialFileReader
```

cbi42 commented:
> Just finished reviewing the improvement to over-charging problem in BackupEngine and left some comments. I personally think it worths a separate PR with its own unit test as well as involving Peter in review cuz it's deeper into BackupEngine stuff that I anticipate.
> 
> See my suggestion here [#9973 (comment)](https://github.com/facebook/rocksdb/pull/9973#discussion_r877804676).
> 
> Thanks for your hard work so far!

Thanks for the detailed review! I've reverted the parts that's worth a separate PR and will come up with unit tests for them later in a separate PR.

cbi42 commented:
> Note: you might want to rebase early cuz the recent merged changes https://github.com/facebook/rocksdb/blob/main/file/sequence_file_reader.cc#L41-L47, https://github.com/facebook/rocksdb/blob/main/utilities/backup/backup_engine_test.cc#L2599-L2748 don't seemed to be included in your PR nor is CI here showing any conflict. This feels weird to me - is your upstream set correctly or fetched recently?
> 
> ```
> $ git rebase --onto upstream/main fddc62a9b459cd8d1eb2dfad8f12217e70ab109a^
> Auto-merging utilities/backup/backup_engine_test.cc
> CONFLICT (content): Merge conflict in utilities/backup/backup_engine_test.cc
> Auto-merging utilities/backup/backup_engine.cc
> Auto-merging util/file_reader_writer_test.cc
> Auto-merging file/sequence_file_reader.cc
> Auto-merging db/repair.cc
> error: could not apply fddc62a9b... Added RateLimiter support in SequentialFileReader
> ```

I've just rebased and fixed all the conflicts, Not sure why CI does not show conflict tho.

cbi42 commented on options/options_parser.cc:
```diff
@@ -271,7 +271,7 @@ Status RocksDBOptionsParser::Parse(const ConfigOptions& config_options_in,
   std::string line;
   // we only support single-lined statement.
-  while (lf_reader.ReadLine(&line)) {
+  while (lf_reader.ReadLine(&line, Env::IO_LOW /* rate_limiter_priority */)) {
```
Updated.

hx235 reviewed (approved):
LGTM! One tiny non-coding tip: an alternative to separating changes into git commits is to "import to FB internal" before each "Ready for review" so that you (and reviewer) can see your incremental changes between each review easily. I found this useful for reviewers, considering git commit's content is mutable by ---amend.
